### ISSUE TITLE
Unable to cast to Chromecast with fling.jar and APP ID

### ISSUE THREAD

**Issue by cksachdev** [LeonNicholls/Fling#20]
I am trying to run fling.jar. And after exploring a bit, I received APPID and launched the fling.jar with APPID as a command line argument. But I am still not able to run any media to Chromecast. Please find below the console log below:

```
Chetans-Macbook-Pro-2:Downloads Apple$ java -jar fling.jar XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
Fling: Fling version 0.6
Fling: Mac OS X
Fling: VLC available
Fling: App ID: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
Fling: set to system default LaF
Fling: createAndShowGUI
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FlingFrame: Started web server on port 8080
Oct 12, 2013 11:41:29 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Starting client on address /239.255.255.250
Oct 12, 2013 11:41:29 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Broadcast client thread starting.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
BroadcastDiscoveryClient: response=HTTP/1.1 200 OK
CACHE-CONTROL: max-age=1800
DATE: Sat, 12 Oct 2013 18:11:35 GMT
EXT:
LOCATION: http://192.168.1.146:8008/ssdp/device-desc.xml
OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
01-NLS: c78422ba-1dd1-11b2-ac45-f41e3cef7c5f
SERVER: Linux/3.8.13, UPnP/1.0, Portable SDK for UPnP devices/1.6.18
X-User-Agent: redsonic
ST: urn:dial-multiscreen-org:service:dial:1
USN: uuid:3db19f46-b3a9-17f4-c127-b17814bae45c::urn:dial-multiscreen-org:service:dial:1
BOOTID.UPNP.ORG: 19
CONFIGID.UPNP.ORG: 1


BroadcastDiscoveryClient: tokens.length=13
FlingFrame: location=http://192.168.1.146:8008/ssdp/device-desc.xml
FlingFrame: appsUrl=http://192.168.1.146:8008/apps/
FlingFrame: modelName=Eureka Dongle
FlingFrame: ChromeCast device found: 192.168.1.146
Oct 12, 2013 11:41:44 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Exiting client loop.
FlingFrame: size=1
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: dragExit event.
FileDrop: normal border restored.
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: drop event.
FileDrop: file list accepted.
DragHereIcon: /Users/Apple/Music/iTunes/iTunes Media/Music/Unknown Artist/Unknown Album/bah phad ke nachan nu.mp3
FlingFrame: sendMediaUrl=/Users/Apple/Music/iTunes/iTunes Media/Music/Unknown Artist/Unknown Album/bah phad ke nachan nu.mp3
FlingFrame: prefix=192.
FlingFrame: interface=en1
FlingFrame: address=/fe80:0:0:0:7ed1:c3ff:fe7e:d1ec%5
FlingFrame: address=/192.168.1.246
FlingFrame: networkAddress=/192.168.1.246
RampClient: device=http://192.168.1.146:8008
RampClient: apps url=http://192.168.1.146:8008/apps/
RampClient: get response code=200
RampClient: lastUrl=http://192.168.1.146:8008/apps/00000000-0000-0000-0000-000000000000
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>00000000-0000-0000-0000-000000000000</name>
  <options allowStop="true"/>
  <state>running</state>
  <link rel="run" href="web-27"/>
  <servicedata xmlns="urn:chrome.google.com:cast">
    <connectionSvcURL>http://192.168.1.146:8008/connection/00000000-0000-0000-0000-000000000000</connectionSvcURL>
    <protocols>
      <protocol>com.google.chromecast.home</protocol>
    </protocols>
  </servicedata>
</service>

RampClient: Content-Type=application/xml
RampClient: Content-Length=530
RampClient: delete response code=200
RampClient: delete response=
RampClient: get response code=404
RampClient: get response=
FileDrop: drop complete.
FileDrop: normal border restored.
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: drop event.
FileDrop: file list accepted.
DragHereIcon: /Users/Apple/Documents/IMG_0194.MOV
FlingFrame: sendMediaUrl=/Users/Apple/Documents/IMG_0194.MOV
FlingFrame: prefix=192.
FlingFrame: interface=en1
FlingFrame: address=/fe80:0:0:0:7ed1:c3ff:fe7e:d1ec%5
FlingFrame: address=/192.168.1.246
FlingFrame: networkAddress=/192.168.1.246
RampClient: get response code=200
RampClient: lastUrl=http://192.168.1.146:8008/apps/00000000-0000-0000-0000-000000000000
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>00000000-0000-0000-0000-000000000000</name>
  <options allowStop="true"/>
  <state>running</state>
  <link rel="run" href="web-28"/>
  <servicedata xmlns="urn:chrome.google.com:cast">
    <connectionSvcURL>http://192.168.1.146:8008/connection/00000000-0000-0000-0000-000000000000</connectionSvcURL>
    <protocols>
      <protocol>com.google.chromecast.home</protocol>
    </protocols>
  </servicedata>
</service>
```

I have downloaded fling.jar from http://goo.gl/HAc9ex and using OSX 10.8.3.

**Comment by cksachdev** [comment]
Few more updates..
After adding "localhost" and "192.168.1.246" ip address in Google Cast extension, Developer Options. The log output looks like below:

```
FileDrop: file list accepted.
DragHereIcon: /Users/Apple/Music/iTunes/iTunes Media/Music/[rKmania.com]/Cocktail (2012)/05 Tera Naam Japdi Phiran - [rKmania.com].mp3
FlingFrame: sendMediaUrl=/Users/Apple/Music/iTunes/iTunes Media/Music/[rKmania.com]/Cocktail (2012)/05 Tera Naam Japdi Phiran - [rKmania.com].mp3
FlingFrame: prefix=192.
FlingFrame: interface=en1
FlingFrame: address=/fe80:0:0:0:7ed1:c3ff:fe7e:d1ec%5
FlingFrame: address=/192.168.1.246
FlingFrame: networkAddress=/192.168.1.246
RampClient: get response code=204
RampClient: device=http://192.168.1.146:8008
RampClient: apps url=http://192.168.1.146:8008/apps/
RampClient: get response code=204
RampClient: get response code=404
RampClient: get response=
FileDrop: drop complete.
FileDrop: normal border restored.
```

Doubt:
Do I need to run Apache or similar server and host an index.html file on my machine?

**Comment by LeonNicholls** [comment]
The receiver html file you provided for whitelisting needs to have your app id.

**Comment by cksachdev** [comment]
I have download index.html from https://github.com/entertailion/Fling/tree/master/receiver and updated this file with my APPID. I hosted this file on MAMP and it is accessible from localhost and 192.168.1.246(my machine's ip). In the whitelisting URL's I had given 192.168.1.246, my website address with a receiver html page ending and 2 more local IP addresses i.e. 192.168.1.xxx. I am still not able to run any media. Here is the updated log output, after these settings:

```
FileDrop: file list accepted.
DragHereIcon: /Users/Apple/Music/iTunes/iTunes Media/Music/DJ BHUVI REMIX/Unknown Album/LAILA TERI - DJ BHUVI REMIX.mp3
FlingFrame: sendMediaUrl=/Users/Apple/Music/iTunes/iTunes Media/Music/DJ BHUVI REMIX/Unknown Album/LAILA TERI - DJ BHUVI REMIX.mp3
FlingFrame: prefix=192.
FlingFrame: interface=en1
FlingFrame: address=/fe80:0:0:0:7ed1:c3ff:fe7e:d1ec%5
FlingFrame: address=/192.168.1.246
FlingFrame: networkAddress=/192.168.1.246
RampClient: get response code=204
RampClient: device=http://192.168.1.146:8008
RampClient: apps url=http://192.168.1.146:8008/apps/
RampClient: get response code=204
RampClient: get response code=404
RampClient: get response=
FileDrop: drop complete.
FileDrop: normal border restored.
```

**Comment by cksachdev** [comment]
Exploring further, I found that fling runs a socket server on the machine. So, I cloned the repo and build it on my machine. Update FlingFrame.java and index.html with my APPID and launched the application. Below is the log output:

```
Chetans-Macbook-Pro-2:Fling Apple$ java -jar target/Fling-0.0.1-SNAPSHOT-jar-with-dependencies.jar 
Fling: Fling version 0.6
Fling: Mac OS X
Fling: VLC available
Fling: set to system default LaF
Fling: createAndShowGUI
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FlingFrame: Started web server on port 8080
Oct 13, 2013 1:43:03 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Starting client on address /239.255.255.250
Oct 13, 2013 1:43:03 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Broadcast client thread starting.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
BroadcastDiscoveryClient: response=HTTP/1.1 200 OK
CACHE-CONTROL: max-age=1800
DATE: Sun, 13 Oct 2013 08:13:09 GMT
EXT:
LOCATION: http://192.168.1.146:8008/ssdp/device-desc.xml
OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
01-NLS: b57dc73c-1dd1-11b2-9aee-f77b5732d992
SERVER: Linux/3.8.13, UPnP/1.0, Portable SDK for UPnP devices/1.6.18
X-User-Agent: redsonic
ST: urn:dial-multiscreen-org:service:dial:1
USN: uuid:3db19f46-b3a9-17f4-c127-b17814bae45c::urn:dial-multiscreen-org:service:dial:1
BOOTID.UPNP.ORG: 21
CONFIGID.UPNP.ORG: 1


BroadcastDiscoveryClient: tokens.length=13
FlingFrame: location=http://192.168.1.146:8008/ssdp/device-desc.xml
FlingFrame: appsUrl=http://192.168.1.146:8008/apps/
FlingFrame: modelName=Eureka Dongle
FlingFrame: ChromeCast device found: 192.168.1.146
Oct 13, 2013 1:43:18 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Exiting client loop.
FlingFrame: size=1
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: dragExit event.
FileDrop: normal border restored.
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: drop event.
FileDrop: file list accepted.
DragHereIcon: /Users/Apple/Music/iTunes/iTunes Media/Music/04 HAAYE MERA DIL/Unknown Album/04 HAAYE MERA DIL.mp3
FlingFrame: sendMediaUrl=/Users/Apple/Music/iTunes/iTunes Media/Music/04 HAAYE MERA DIL/Unknown Album/04 HAAYE MERA DIL.mp3
FlingFrame: prefix=192.
FlingFrame: interface=en1
FlingFrame: address=/fe80:0:0:0:7ed1:c3ff:fe7e:d1ec%5
FlingFrame: address=/192.168.1.246
FlingFrame: networkAddress=/192.168.1.246
RampClient: device=http://192.168.1.146:8008
RampClient: apps url=http://192.168.1.146:8008/apps/
RampClient: get response code=200
RampClient: lastUrl=http://192.168.1.146:8008/apps/00000000-0000-0000-0000-000000000000
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>00000000-0000-0000-0000-000000000000</name>
  <options allowStop="true"/>
  <state>running</state>
  <link rel="run" href="web-15"/>
  <servicedata xmlns="urn:chrome.google.com:cast">
    <connectionSvcURL>http://192.168.1.146:8008/connection/00000000-0000-0000-0000-000000000000</connectionSvcURL>
    <protocols>
      <protocol>com.google.chromecast.home</protocol>
    </protocols>
  </servicedata>
</service>

RampClient: Content-Type=application/xml
RampClient: Content-Length=530
RampClient: delete response code=200
RampClient: delete response=
RampClient: get response code=404
RampClient: get response=
FileDrop: drop complete.
FileDrop: normal border restored.
RampClient: get response code=200
RampClient: lastUrl=http://192.168.1.146:8008/apps/00000000-0000-0000-0000-000000000000
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>00000000-0000-0000-0000-000000000000</name>
  <options allowStop="true"/>
  <state>running</state>
  <link rel="run" href="web-16"/>
  <servicedata xmlns="urn:chrome.google.com:cast">
    <connectionSvcURL>http://192.168.1.146:8008/connection/00000000-0000-0000-0000-000000000000</connectionSvcURL>
    <protocols>
      <protocol>com.google.chromecast.home</protocol>
    </protocols>
  </servicedata>
</service>

RampClient: Content-Type=application/xml
RampClient: Content-Length=530
RampClient: delete response code=200
RampClient: delete response=
Chetans-Macbook-Pro-2:Fling Apple$ 
```

Can you help me, what is the step I am missing here?

I tried to access http://192.168.1.246:8080/ from browser, and it shows this message:
FORBIDDEN: Reading file failed.

**Comment by LeonNicholls** [comment]
Make sure you have configured your Chromecast device to send its serial number to Google. Also, restart the Chromecast device to make sure it is synced with the Google servers.

**Comment by cksachdev** [comment]
How do I configure Chromecast device to send the serial number?

**Comment by LeonNicholls** [comment]
https://developers.google.com/cast/whitelisting

**Comment by cksachdev** [comment]
Done. Still no success. Is there a way to validate that it is Synced with Google servers? Rebooted multiple times also validated that the checkbox to sync is checked in Chromecast app. When I drag an item to the Fling app, the Chromecast icon in TV runs its animation.
localhost:8080 in firefox, actually runs the media, which I dragged, while in Chrome the same URL gives an option to download.

**Comment by cksachdev** [comment]
Revising all the steps, I found that I was giving wrong APPID for the current IP address. After giving the right APPID, there was a different message in TV. It was in brain freeze mode. with a message, something could not load. Below is a log message for this entire action

```
Chetans-Macbook-Pro-2:Fling Apple$ java -jar target/Fling-0.0.1-SNAPSHOT-jar-with-dependencies.jar 
Fling: Fling version 0.6
Fling: Mac OS X
Fling: VLC available
Fling: set to system default LaF
Fling: createAndShowGUI
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FlingFrame: Started web server on port 8080
Oct 13, 2013 11:19:03 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Starting client on address /239.255.255.250
Oct 13, 2013 11:19:03 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Broadcast client thread starting.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
FileDrop: Hierarchy changed.
FileDrop: Drop target added to component.
BroadcastDiscoveryClient: response=HTTP/1.1 200 OK
CACHE-CONTROL: max-age=1800
DATE: Sun, 13 Oct 2013 17:49:03 GMT
EXT:
LOCATION: http://192.168.1.146:8008/ssdp/device-desc.xml
OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
01-NLS: 12c42d16-1dd2-11b2-9567-8305880307cc
SERVER: Linux/3.8.13, UPnP/1.0, Portable SDK for UPnP devices/1.6.18
X-User-Agent: redsonic
ST: urn:dial-multiscreen-org:service:dial:1
USN: uuid:3db19f46-b3a9-17f4-c127-b17814bae45c::urn:dial-multiscreen-org:service:dial:1
BOOTID.UPNP.ORG: 26
CONFIGID.UPNP.ORG: 1


BroadcastDiscoveryClient: tokens.length=13
FlingFrame: location=http://192.168.1.146:8008/ssdp/device-desc.xml
FlingFrame: appsUrl=http://192.168.1.146:8008/apps/
FlingFrame: modelName=Eureka Dongle
FlingFrame: ChromeCast device found: 192.168.1.146
Oct 13, 2013 11:19:18 PM com.entertailion.java.fling.Log i
INFO: BroadcastDiscoveryClient: Exiting client loop.
FlingFrame: size=1
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: dragExit event.
FileDrop: normal border restored.
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: dragExit event.
FileDrop: normal border restored.
FileDrop: dragEnter event.
FileDrop: normal border saved.
FileDrop: drag border set.
FileDrop: event accepted.
FileDrop: drop event.
FileDrop: file list accepted.
DragHereIcon: /Users/Apple/Movies/Jaksta/Converted/bah phad ke nachan nu.mp3
FlingFrame: sendMediaUrl=/Users/Apple/Movies/Jaksta/Converted/bah phad ke nachan nu.mp3
FlingFrame: prefix=192.
FlingFrame: interface=en1
FlingFrame: address=/fe80:0:0:0:7ed1:c3ff:fe7e:d1ec%5
FlingFrame: address=/192.168.1.246
FlingFrame: networkAddress=/192.168.1.246
RampClient: device=http://192.168.1.146:8008
RampClient: apps url=http://192.168.1.146:8008/apps/
RampClient: get response code=200
RampClient: lastUrl=http://192.168.1.146:8008/apps/00000000-0000-0000-0000-000000000000
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>00000000-0000-0000-0000-000000000000</name>
  <options allowStop="true"/>
  <state>running</state>
  <link rel="run" href="web-19"/>
  <servicedata xmlns="urn:chrome.google.com:cast">
    <connectionSvcURL>http://192.168.1.146:8008/connection/00000000-0000-0000-0000-000000000000</connectionSvcURL>
    <protocols>
      <protocol>com.google.chromecast.home</protocol>
    </protocols>
  </servicedata>
</service>

RampClient: Content-Type=application/xml
RampClient: Content-Length=530
RampClient: delete response code=200
RampClient: delete response=
RampClient: get response code=200
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>5df13271-7add-4c34-8e3b-ef01f56752cb_1</name>
  <options allowStop="true"/>
  <state>stopped</state>
</service>

RampClient: state=stopped
RampClient: post response code=201
RampClient: post response=
RampClient: post response location=http://192.168.1.146:8008/apps/5df13271-7add-4c34-8e3b-ef01f56752cb_1/web-21
RampClient: Access-Control-Allow-Methods=GET, POST, DELETE, OPTIONS
RampClient: Access-Control-Expose-Headers=Location
RampClient: Location=http://192.168.1.146:8008/apps/5df13271-7add-4c34-8e3b-ef01f56752cb_1/web-21
RampClient: Access-Control-Allow-Origin=chrome-extension://boadgeojelhgndaghljhdicfkmllpafd
RampClient: Content-Length=0
RampClient: get response code=200
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>5df13271-7add-4c34-8e3b-ef01f56752cb_1</name>
  <options allowStop="true"/>
  <state>running</state>
  <link rel="run" href="web-21"/>
  <servicedata xmlns="urn:chrome.google.com:cast">
  </servicedata>
</service>

RampClient: state=running
RampClient: connectionServiceUrl=null
RampClient: protocol=null
RampClient: get response code=200
RampClient: get response=<?xml version="1.0" encoding="UTF-8"?>
<service xmlns="urn:dial-multiscreen-org:schemas:dial">
  <name>5df13271-7add-4c34-8e3b-ef01f56752cb_1</name>
  <options allowStop="true"/>
  <state>stopped</state>
</service>

RampClient: state=stopped
RampClient: connectionServiceUrl=null
RampClient: protocol=null
Oct 13, 2013 11:19:31 PM com.entertailion.java.fling.Log i
INFO: RampClient: connectionServiceUrl is null
FileDrop: drop complete.
FileDrop: normal border restored.

```

**Comment by cksachdev** [comment]
Tried following extensions:
flv, mp3, mp4, mov. Doesn't seem to work with any file formats. Any suggestions?

What is connectionServiceUrl?

Another thing noticed in Fling app is, I drag any media format, it doesn't gets the duration. Is it supposed to show the duration? I noticed in the index.html, there is checkStatus function, which depends on media time?

**Comment by LeonNicholls** [comment]
Confirm that you can load the whitelisted url for your app id on your local network.

**Comment by cksachdev** [comment]
Got it working. Here is a description of the problem:
For whitelisting I had given 192.168.1.46 IP Address, so index.html with APPID should be available at 192.168.1.46/index.html With fling, it was running 192.168.1.146:8080, so somehow it was not working. Once I launched MAMP and got index.html at 192.168.1.146, it started working :) 

Thx
